File: data_maintenance_tools/data_entry/location_parse.py
"""Parse city and US state from Metal Archives and MusicBrainz location data."""

from __future__ import annotations

from typing import Any

US_STATE_NAME_TO_CODE: dict[str, str] = {
    "alabama": "AL",
    "alaska": "AK",
    "arizona": "AZ",
    "arkansas": "AR",
    "california": "CA",
    "colorado": "CO",
    "connecticut": "CT",
    "delaware": "DE",
    "district of columbia": "DC",
    "florida": "FL",
    "georgia": "GA",
    "hawaii": "HI",
    "idaho": "ID",
    "illinois": "IL",
    "indiana": "IN",
    "iowa": "IA",
    "kansas": "KS",
    "kentucky": "KY",
    "louisiana": "LA",
    "maine": "ME",
    "maryland": "MD",
    "massachusetts": "MA",
    "michigan": "MI",
    "minnesota": "MN",
    "mississippi": "MS",
    "missouri": "MO",
    "montana": "MT",
    "nebraska": "NE",
    "nevada": "NV",
    "new hampshire": "NH",
    "new jersey": "NJ",
    "new mexico": "NM",
    "new york": "NY",
    "north carolina": "NC",
    "north dakota": "ND",
    "ohio": "OH",
    "oklahoma": "OK",
    "oregon": "OR",
    "pennsylvania": "PA",
    "rhode island": "RI",
    "south carolina": "SC",
    "south dakota": "SD",
    "tennessee": "TN",
    "texas": "TX",
    "utah": "UT",
    "vermont": "VT",
    "virginia": "VA",
    "washington": "WA",
    "west virginia": "WV",
    "wisconsin": "WI",
    "wyoming": "WY",
}

_COUNTRY_TOKENS = frozenset(
    {
        "united states",
        "usa",
        "us",
        "u.s.",
        "u.s.a.",
        "united kingdom",
        "uk",
        "u.k.",
        "england",
        "scotland",
        "wales",
        "northern ireland",
    }
)


def state_name_to_code(value: str) -> str:
    """Convert a US state name or existing abbreviation to a two-letter code."""
    raw = _normalize_token(value)
    if not raw:
        return ""
    if len(raw) == 2 and raw.isalpha():
        return raw.upper()
    compact = raw.replace(".", "")
    if len(compact) == 2 and compact.isalpha():
        return compact.upper()
    return US_STATE_NAME_TO_CODE.get(raw.lower(), "")


def _normalize_token(value: str) -> str:
    return (value or "").strip().rstrip(".")


def _is_united_states(country: str) -> bool:
    normalized = (country or "").strip().lower()
    return normalized in {"united states", "usa", "us", "u.s.", "u.s.a."}


def _allow_us_state_in_location(country: str) -> bool:
    """Whether a comma-separated location may contain a US state name."""
    normalized = (country or "").strip().lower()
    if not normalized or _is_united_states(country):
        return True
    return False


def _state_from_segment(segment: str) -> str:
    return state_name_to_code(segment)


def parse_ma_location(location: str, country: str = "") -> tuple[str, str]:
    """
    Parse Metal Archives Location text into city and optional US state code.

    MA stores full US state names (e.g. "Atlanta, Georgia"). The state name is
    translated to a two-letter code whenever it is recognized.
    """
    location = (location or "").strip()
    if not location:
        return "", ""

    parts = [_normalize_token(part) for part in location.split(",") if _normalize_token(part)]
    if not parts:
        return "", ""

    if _allow_us_state_in_location(country):
        state_index = -1
        state_code = ""
        for index in range(len(parts) - 1, -1, -1):
            token = parts[index].lower()
            if token in _COUNTRY_TOKENS:
                continue
            code = _state_from_segment(parts[index])
            if code:
                state_code = code
                state_index = index
                break

        if state_code:
            city = ", ".join(parts[:state_index]).strip() if state_index > 0 else ""
            return city, state_code

        if len(parts) == 1:
            code = _state_from_segment(parts[0])
            if code:
                return "", code

    return parts[0], ""


def parse_musicbrainz_location(detail: dict[str, Any]) -> tuple[str, str]:
    """Extract city and US state code from a MusicBrainz artist record."""
    country_code = (detail.get("country") or "").strip().upper()
    begin_area = detail.get("begin-area") or {}
    area = detail.get("area") or {}

    city = (begin_area.get("name") or "").strip()
    state = ""

    area_type = (area.get("type") or "").strip().lower()
    area_name = (area.get("name") or "").strip()

    if country_code == "US":
        if area_type == "subdivision" and area_name:
            state = state_name_to_code(area_name)
        elif not city and area_type in {"city", "municipality", "town"} and area_name:
            city = area_name
    elif not city and area_name and area_type in {"city", "municipality", "town"}:
        city = area_name

    return city, state

File: promoter_admin/lib/src/services/location_parse.dart
/// Parse city and US state from Metal Archives and MusicBrainz location data.

const _usStateNameToCode = <String, String>{
  'alabama': 'AL',
  'alaska': 'AK',
  'arizona': 'AZ',
  'arkansas': 'AR',
  'california': 'CA',
  'colorado': 'CO',
  'connecticut': 'CT',
  'delaware': 'DE',
  'district of columbia': 'DC',
  'florida': 'FL',
  'georgia': 'GA',
  'hawaii': 'HI',
  'idaho': 'ID',
  'illinois': 'IL',
  'indiana': 'IN',
  'iowa': 'IA',
  'kansas': 'KS',
  'kentucky': 'KY',
  'louisiana': 'LA',
  'maine': 'ME',
  'maryland': 'MD',
  'massachusetts': 'MA',
  'michigan': 'MI',
  'minnesota': 'MN',
  'mississippi': 'MS',
  'missouri': 'MO',
  'montana': 'MT',
  'nebraska': 'NE',
  'nevada': 'NV',
  'new hampshire': 'NH',
  'new jersey': 'NJ',
  'new mexico': 'NM',
  'new york': 'NY',
  'north carolina': 'NC',
  'north dakota': 'ND',
  'ohio': 'OH',
  'oklahoma': 'OK',
  'oregon': 'OR',
  'pennsylvania': 'PA',
  'rhode island': 'RI',
  'south carolina': 'SC',
  'south dakota': 'SD',
  'tennessee': 'TN',
  'texas': 'TX',
  'utah': 'UT',
  'vermont': 'VT',
  'virginia': 'VA',
  'washington': 'WA',
  'west virginia': 'WV',
  'wisconsin': 'WI',
  'wyoming': 'WY',
};

const _countryTokens = {
  'united states',
  'usa',
  'us',
  'u.s.',
  'u.s.a.',
  'united kingdom',
  'uk',
  'u.k.',
  'england',
  'scotland',
  'wales',
  'northern ireland',
};

/// Convert a US state name or existing abbreviation to a two-letter code.
String stateNameToCode(String value) {
  final raw = _normalizeToken(value);
  if (raw.isEmpty) return '';
  if (raw.length == 2 && RegExp(r'^[A-Za-z]+$').hasMatch(raw)) {
    return raw.toUpperCase();
  }
  final compact = raw.replaceAll('.', '');
  if (compact.length == 2 && RegExp(r'^[A-Za-z]+$').hasMatch(compact)) {
    return compact.toUpperCase();
  }
  return _usStateNameToCode[raw.toLowerCase()] ?? '';
}

String _normalizeToken(String value) => value.trim().replaceFirst(RegExp(r'\.$'), '');

bool _isUnitedStates(String country) {
  final normalized = country.trim().toLowerCase();
  return normalized == 'united states' ||
      normalized == 'usa' ||
      normalized == 'us' ||
      normalized == 'u.s.' ||
      normalized == 'u.s.a.';
}

bool _allowUsStateInLocation(String country) {
  final normalized = country.trim().toLowerCase();
  if (normalized.isEmpty || _isUnitedStates(country)) return true;
  return false;
}

/// Parse Metal Archives Location text into city and optional US state code.
({String city, String state}) parseMaLocation(
  String location, {
  String country = '',
}) {
  final trimmed = location.trim();
  if (trimmed.isEmpty) return (city: '', state: '');

  final parts = trimmed
      .split(',')
      .map(_normalizeToken)
      .where((p) => p.isNotEmpty)
      .toList();
  if (parts.isEmpty) return (city: '', state: '');

  if (_allowUsStateInLocation(country)) {
    var stateIndex = -1;
    var stateCode = '';
    for (var index = parts.length - 1; index >= 0; index--) {
      final token = parts[index].toLowerCase();
      if (_countryTokens.contains(token)) continue;
      final code = stateNameToCode(parts[index]);
      if (code.isNotEmpty) {
        stateCode = code;
        stateIndex = index;
        break;
      }
    }

    if (stateCode.isNotEmpty) {
      final city =
          stateIndex > 0 ? parts.sublist(0, stateIndex).join(', ').trim() : '';
      return (city: city, state: stateCode);
    }

    if (parts.length == 1) {
      final code = stateNameToCode(parts[0]);
      if (code.isNotEmpty) return (city: '', state: code);
    }
  }

  return (city: parts[0], state: '');
}

/// Extract city and US state code from a MusicBrainz artist record.
({String city, String state}) parseMusicBrainzLocation(
  Map<String, dynamic> detail,
) {
  final countryCode = (detail['country']?.toString() ?? '').trim().toUpperCase();
  final beginArea = detail['begin-area'];
  final area = detail['area'];
  final beginMap =
      beginArea is Map ? Map<String, dynamic>.from(beginArea) : <String, dynamic>{};
  final areaMap =
      area is Map ? Map<String, dynamic>.from(area) : <String, dynamic>{};

  var city = (beginMap['name']?.toString() ?? '').trim();
  var state = '';

  final areaType = (areaMap['type']?.toString() ?? '').trim().toLowerCase();
  final areaName = (areaMap['name']?.toString() ?? '').trim();

  if (countryCode == 'US') {
    if (areaType == 'subdivision' && areaName.isNotEmpty) {
      state = stateNameToCode(areaName);
    } else if (city.isEmpty &&
        {'city', 'municipality', 'town'}.contains(areaType) &&
        areaName.isNotEmpty) {
      city = areaName;
    }
  } else if (city.isEmpty &&
      areaName.isNotEmpty &&
      {'city', 'municipality', 'town'}.contains(areaType)) {
    city = areaName;
  }

  return (city: city, state: state);
}
